packages/contracts/contracts/adapters/UniswapV3Adapter.sol

```diff
@@ -0,0 +1,122 @@
+    /// @param token Token address
+    /// @param permission Whether token is allowed
+    function setTokenPermission(
+        address token,
+        bool permission
+    ) external onlyOwner {
+        _allowedTokens[token] = permission;
+        emit TokenPermissionSet(token, permission);
+    }
+
+    /// @notice Proxy for exactInputSingle function of Uniswap SwapRouter. Ensures tokenIn
+    ///         is allowed and recipient is caller.
+    /// @param params ExactInputSingleParams (see ISwapRouter)
+    function exactInputSingle(
+        ISwapRouter.ExactInputSingleParams memory params
+    ) external returns (uint256 amountOut) {
+        require(_allowedTokens[params.tokenIn], "!allowed token");
+        require(params.recipient == msg.sender, "!recipient");
+
+        IERC20(params.tokenIn).transferFrom(
+            address(msg.sender),
+            address(this),
+            params.amountIn
+        );
+        IERC20(params.tokenIn).approve(address(_swapRouter), params.amountIn);
+
+        amountOut = _swapRouter.exactInputSingle(params);
+    }
+
+    /// @notice Proxy for exactInput function of Uniswap SwapRouter.
+    ///         Ensures all tokens in path are is allowed and recipient is caller.
+    /// @param params ExactInputParams (see ISwapRouter)
+    function exactInput(
+        ISwapRouter.ExactInputParams memory params
+    ) external returns (uint256 amountOut) {
+        require(tokensAreAllowed(params.path), "!allowed path");
+        require(params.recipient == msg.sender, "!recipient");
+
+        address tokenIn = extractTokenAddressFromPath(params.path, 0);
+
+        IERC20(tokenIn).transferFrom(
+            address(msg.sender),
+            address(this),
+            params.amountIn
+        );
+        IERC20(tokenIn).approve(address(_swapRouter), params.amountIn);
+
+        amountOut = _swapRouter.exactInput(params);
+    }
+
+    /// @notice Checks that all tokens in path are allowed, returns false if any are not
+    /// @param path Path of tokens
+    /// @dev path is formatted as [address token1][uint24 fee][address token2], thus the length will
+    ///      always be ((20 + 3) * n) + 20 bytes where n is the number of hops
+    function tokensAreAllowed(bytes memory path) internal view returns (bool) {
+        uint256 length = path.length;
+
+        // Ensure the path length is correct: ((20 + 3) * n) + 20
+        require((length - 20) % 23 == 0, "Invalid path length");
+
+        uint256 numAddresses = (length - 20) / 23 + 1;
+
+        for (uint256 i = 0; i < numAddresses; i++) {
+            address tokenAddress = extractTokenAddressFromPath(path, i * 23);
+            if (!_allowedTokens[tokenAddress]) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /// @notice Extracts token address from path at a given index into the path
+    /// @param path Path of tokens
+    /// @param index Index into path
```

Comment:
This is a bit misleading, because `index` isn't actually the `index`, it's the offset of the `index`th `(fee, tokenAddr)` pair in the path, less the length prefix.

Instead, lets keep all of the pointer arithmetic in one place and move the mul by 23 in here

```
// Load the `index`th token in the path.
// Since each `(fee, tokenAddr)` pair in the encoded path is 23 bytes,
// we get the 32-byte value at offset `32 (length prefix) + 23 * i`
// and then shift right by 12 bytes to get just the first 20 bytes of it
assembly {
    tokenAddr := shr(96, mload(add(add(path, 0x20), mul(0x17, index))))
}
```

Then the caller just passes in `i` to this function.